ddl/ddl_api.go

```diff
@@ -3549,9 +3559,77 @@ func (d *ddl) CreatePrimaryKey(ctx sessionctx.Context, ti ast.Ident, indexName m
 	return errors.Trace(err)
 }
 
-func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.IndexKeyType, indexName model.CIStr,
-	idxColNames []*ast.IndexPartSpecification, indexOption *ast.IndexOption, ifNotExists bool) error {
+func buildHiddenColumnInfo(ctx sessionctx.Context, t table.Table, indexPartSpecifications []*ast.IndexPartSpecification, indexName model.CIStr) ([]*model.ColumnInfo, error) {
+	tblInfo := t.Meta()
+	hiddenCols := make([]*model.ColumnInfo, 0, len(indexPartSpecifications))
+	for i, idxPart := range indexPartSpecifications {
+		if idxPart.Expr == nil {
+			continue
+		}
+		idxPart.Column = &ast.ColumnName{Name: model.NewCIStr(fmt.Sprintf("%s_%s_%d", expressionIndexPrefix, indexName, i))}
+		// Check whether the hidden columns have existed.
+		col := table.FindCol(t.Cols(), idxPart.Column.Name.L)
+		if col != nil {
+			// TODO: Use expression index related error.
+			return nil, infoschema.ErrColumnExists.GenWithStackByArgs(col.Name.String())
+		}
+		idxPart.Length = types.UnspecifiedLength
+		// The index part is an expression, prepare a hidden column for it.
+		if len(idxPart.Column.Name.L) > mysql.MaxColumnNameLength {
+			// TODO: Refine the error message.
+			return nil, ErrTooLongIdent.GenWithStackByArgs("hidden column")
+		}
+		// TODO: refine the error message.
+		if err := checkIllegalFn4GeneratedColumn("expression index", idxPart.Expr); err != nil {
+			return nil, errors.Trace(err)
+		}
 
+		var sb strings.Builder
+		restoreFlags := format.RestoreStringSingleQuotes | format.RestoreKeyWordLowercase | format.RestoreNameBackQuotes |
+			format.RestoreSpacesAroundBinaryOperation
+		restoreCtx := format.NewRestoreCtx(restoreFlags, &sb)
+		sb.Reset()
+		err := idxPart.Expr.Restore(restoreCtx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		expr, err := expression.RewriteSimpleExprWithTableInfo(ctx, tblInfo, idxPart.Expr)
+		if err != nil {
+			// TODO: refine the error message.
+			return nil, err
+		}
+		if _, ok := expr.(*expression.Column); ok {
+			return nil, ErrFunctionalIndexOnField
+		}
+
+		colInfo := &model.ColumnInfo{
+			Name:                idxPart.Column.Name,
+			GeneratedExprString: sb.String(),
+			GeneratedStored:     false,
+			Version:             model.CurrLatestColumnInfoVersion,
+			Dependences:         make(map[string]struct{}),
+			Hidden:              true,
```

Comment:
The method of `WritableCols` should ignore this hidden column too?

Reply:
No, for `update`, we need to write the hidden column.